### PR TITLE
einsum using cublas

### LINKED ISSUE #1197: Intermediate array of cupy.einsum is unnecessarily large
The current implementation of `cupy.einsum` has much different time complexity than `numpy.einsum` with `optimize=False`.  For example, `cupy.einsum('i,i->', x, y)` is O(n^2)-time, because it computes the diagonal of the tensor product of `x` and `y`.
https://github.com/cupy/cupy/blob/v5.0.0a1/cupy/linalg/einsum.py#L342

### LINKED ISSUE #1199: cupy.einsum does not return a writeable view
numpy has this feature from 1.10.  This feature is not supported yet, while the name of the internal function `calc_single_view` contains "view".

To reproduce:
```
>>> import cupy
>>> a = cupy.arange(6).reshape(2, 3)
>>> cupy.einsum('ij', a)[...] = 1
>>> a
array([[0, 1, 2],
       [3, 4, 5]])
>>> import numpy
>>> a = numpy.arange(6).reshape(2, 3)
>>> numpy.einsum('ij', a)[...] = 1
>>> a
array([[1, 1, 1],
       [1, 1, 1]])
```
(cupy==5.0.0a1)

### LINKED ISSUE #1214: cupy.einsum does not support broadcast along non-ellipsis axes
I don't know why numpy supports this.
```
>>> cupy.einsum('i,i', cupy.ones(3), cupy.ones(1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kataoka/cupy/cupy/linalg/einsum.py", line 343, in einsum
    result, subscript = calc_single_view(result, subscript)
  File "/home/kataoka/cupy/cupy/linalg/einsum.py", line 56, in calc_single_view
    shape_b))
ValueError: dimensions in operand 0 for collapsing index 'i' don't match (1 != 3)
>>> numpy.einsum('i,i', numpy.ones(3), numpy.ones(1))
3.0
```

BTW, diagonal axes should match.  (Related: #1197)
```
>>> numpy.einsum('ii', numpy.ones((3, 1)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kataoka/.pyenv/versions/anaconda3-4.2.0/envs/n/lib/python3.6/site-packages/numpy/core/einsumfunc.py", line 1069, in einsum
    return c_einsum(*operands, **kwargs)
ValueError: dimensions in operand 0 for collapsing index 'i' don't match (3 != 1)
```

### PR DESCRIPTION
**Pull request by toslunar**
Close #1197.

Method: Call `<T>gemmStridedBatched` via `cupy.matmul`.

This PR also closes #1199, and closes #1214.

### CODE REVIEW

File: tests/cupy_tests/linalg_tests/test_einsum.py [hunk 1]
```diff
@@ -1,10 +1,44 @@
...
+    return tuple(1 if s == 1 else max(0, s - dec) for s in shape)
+
+
+def _rand1_shape(shape, prob):
+    # return tuple(1 if numpy.random.rand() < prob else s for s in shape)
+    table = {}
+    new_shape = []
+    for s in shape:
+        if s not in table:
+            table[s] = 1 if numpy.random.rand() < prob else s
```
Comment (toslunar) [inline tests/cupy_tests/linalg_tests/test_einsum.py]: In order to close #1214

File: tests/cupy_tests/linalg_tests/test_einsum.py [hunk 2]
```diff
@@ -153,21 +252,57 @@ def test_invalid_arrow4(self, xp):
...
+        out = xp.einsum(self.subscripts, a)
+        if xp is not numpy:
+            optimized_out = xp.einsum(self.subscripts, a, optimize=True)
+            testing.assert_allclose(optimized_out, out)
+        return out
+
+    @testing.with_requires('numpy>=1.10')
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_equal()
+    def test_einsum_unary_views(self, xp, dtype):
```
Comment (toslunar) [inline tests/cupy_tests/linalg_tests/test_einsum.py]: In order to close #1199

**Comment by toslunar** [comment]
I pasted my git history into this PR for the readability.

**Comment by mscipio** [comment]
Any plan on merging this in the near future? As shown in #1275 (and maybe somewhere else, too), at the moment cupy.einsum is there but it is not really useful if you can use it only for toy matrices ...

File: cupy/linalg/einsum.py [hunk 3]
```diff
@@ -1,212 +1,323 @@
...
+
+
+def _chr(char):
+    if char < 0:
+        return "...[%d]" % char
+    else:
+        return chr(char)
+
+
+def _parse_ellipsis_subscript(subscript, ndim=None, ellipsis_len=None):
```
Comment (asi1024) [inline cupy/linalg/einsum.py]: Could you write a comment that only either of the kwargs required to be `None`?

File: cupy/linalg/einsum.py [hunk 4]
```diff
@@ -1,212 +1,323 @@
...
 
-    assert len(set(output_subscript)) == len(output_subscript)
-    assert set(input_subscript) == set(output_subscript)
+    if len(operands) == 0:
+        raise ValueError("No input operands")
 
-    if input_subscript == output_subscript:
-        return ioperand
+    if isinstance(operands[0], str):
+        subscripts = operands[0].replace(" ", "")
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/linalg/einsum.py]: Could you add such test cases?

File: cupy/linalg/einsum.py [hunk 5]
```diff
@@ -231,119 +369,207 @@ def einsum(*operands):
...
+    ]
+
+    # Get length of each unique dimension and ensure all dimensions are correct
+    dimension_dict = {}
+    for tnum, term in enumerate(input_subscripts):
+        sh = operands[tnum].shape
+        for cnum, char in enumerate(term):
+            dim = sh[cnum]
+            if char in dimension_dict.keys():
+                # For broadcasting cases we always want the largest dim size
```
Comment (asi1024) [inline cupy/linalg/einsum.py]: Don't allow broadcasting by default.

The following code fails only when `xp=numpy` and `opt=True`.

```
import numpy
import cupy

for xp in (cupy, numpy):
    for opt in (False, True):
        res = xp.einsum('i,i->i', xp.array([1, 2]), xp.array([1]), optimize=opt)
        print(res)
```

Comment (toslunar) [reply cupy/linalg/einsum.py]: I'll make the test require `numpy!=1.14.0`.

**Comment by toslunar** [comment]
Fixed

File: cupy/linalg/einsum.py [hunk 6]
```diff
@@ -231,119 +371,200 @@ def einsum(*operands):
...
+        if not options['sum_ellipsis']:
+            if '@' not in output_subscript and -1 in dimension_dict:
+                raise ValueError(
+                    "output has more dimensions than subscripts "
+                    "given in einstein sum, but no '...' ellipsis "
+                    "provided to broadcast the extra dimensions.")
+        output_subscript = _parse_ellipsis_subscript(
+            output_subscript, None,
+            ellipsis_len=len(
+                [label for label in dimension_dict.keys() if label < 0])
```
Comment (asi1024) [inline cupy/linalg/einsum.py]: `ellipsis_len=sum(label < 0 for label in dimension_dict.keys()`

File: cupy/linalg/einsum.py [hunk 7]
```diff
@@ -231,119 +371,200 @@ def einsum(*operands):
...
+                squeeze_indices = []
+                sub = []
+                for axis, label in enumerate(input_subscripts[idx]):
+                    if arr.shape[axis] == 1:
+                        squeeze_indices.append(axis)
+                    else:
+                        sub.append(label)
+                input_subscripts[idx] = sub
+                operands[idx] = cupy.squeeze(arr, axis=tuple(squeeze_indices))
+                assert len(operands[idx].shape) == len(input_subscripts[idx])
```
Comment (asi1024) [inline cupy/linalg/einsum.py]: How about using `operands[idx].ndim`?

File: cupy/linalg/einsum.py [hunk 8]
```diff
@@ -231,119 +371,200 @@ def einsum(*operands):
...
+            del arr
+
+    # unary einsum without summation should return a (writeable) view
+    returns_view = len(operands) == 1
+
+    # unary sum
+    for idx, sub in enumerate(input_subscripts):
+        other_subscripts = list(input_subscripts)
```
Comment (asi1024) [inline cupy/linalg/einsum.py]: This `input_subscripts` is guaranteed to be a list object.
Use `copy.copy` instead of `list()`.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**